### DISCUSSION TITLE
Path forward for deprecating ROPC/Direct Access Grants in Keycloak?

### DISCUSSION THREAD

**Discussion by jsorah** [keycloak/keycloak#24585] [Ideas]
# The Problem
Today, Keycloak supports ROPC flow as mentioned in the OAuth2 specification. It would be nice to see Keycloak deprecate this functionality and guide users against using it, and even in some cases make it so that it could be turned off entirely if desired. [OAuth 2.0 Security Best Current Practice](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-24#name-resource-owner-password-cre) states very clearly that ROPC should not be used, and I often wonder if the user base is aware of that fact.  There are also some folks asking about OAuth 2.1 support in Keycloak, which removes both ROPC and implicit flows entirely.

# Proposal
There are a few things I feel could be done to move Keycloak forward on this front and I wonder what else the maintainers and community thinks we could do to guide users away from using ROPC.

1) Stop setting "Direct Access Grants Enabled" checked by default on the Admin UI when adding a new OIDC client. :heavy_check_mark:  Done via - [Issue 30226](https://github.com/keycloak/keycloak/issues/30226)

2) Adding a warning symbol or similar callout on the Admin UI that this is not a recommended setting when it is used.

3) Feature flagging ROPC as a feature - disabling functionality and no longer advertising on the OIDC Well Known Endpoint as a supported grant type. This would allow users to opt-out of ROPC in their deployments altogether. Client policies could fit here in some manner, but it would feel better if there was no possibility of the flow being used at all because the code path was explicitly off. Client policies also does not provide the ability to keep clients from discovering the grant type being supported.

4) Eventually moving ROPC feature as "off" by default

5) ~~Removing ROPC outright from the code base~~ - :x: - Not likely to happen given MANY long lived edge cases and legacy needs still fulfilled by ROPC.

## Some considerations
I can see some quick, positive gains with 1-3 but 4-5 are certainly longer term items.  Item 5 in particular has some other implications as I think some of the integration tests today rely on some level of ROPC to generate tokens.  I think there are also items around `kcadm.sh` type scripts where support for device auth or some native app authorization code flow could be added as an authentication option to replace ROPC where service accounts would not fit the use case?

**Comment by hedgelot** [comment]
I am curious what Keycloaks plans are on direct access grants in the future. 
We always try to go with the code flow or the device flow to login in our applications. But sometimes due to technical restrictions on the devices and functional requirement to not use external devices, we can't use the browser at all. In that case our only option is to use ROPC. In our case it's always first-party apps (we develop the apps and manage Keycloak), so username/password login should be OK. OAuth2 seems to be focused on login for 3rd party clients where ROPC is bad practice, but there's no 1st party alternative in Keycloak as far as I know.

**Reply by mbraeudev** [reply]
It might be interesting to have a look at this draft then: https://datatracker.ietf.org/doc/draft-ietf-oauth-first-party-apps/
> This document, OAuth for First-Party Apps (FiPA), extends the OAuth  2.0 Authorization Framework [RFC6749] with a new endpoint, authorization_challenge_endpoint, to support first-party applications that want to control the process of obtaining authorization from the user using a native experience.

There is already an ongoing discussion here: https://github.com/keycloak/keycloak/discussions/25014

**Comment by Elektordi** [comment]
Maybe points 1 and 2 could be implemented easily without any impact on existing configurations.
On our in-house Keycloak, we have a patch (more like a "hack") in the build process to disable this checkbox by default.
Do you think it is ready to create a PR on those first points or does it needs more validation before?

**Comment by ahahu** [comment]
I'm strongly advocating for NOT removing Direct Access Grants because we heavily depend on it to authenticate service accounts that use our REST services with Basic Authentication. Moving the complete customer base (mostly financial sector) to another mechanism is a journey that would take several years.

**Comment by stianst** [comment]
I agree that ROPC should really suffer a terrible death; it makes little sense as a way to interact with an IdP, especially these days.

That being said I also fully appreciate that many have decided to use it regardless, hence removing it completely is not something that will happen anytime soon.

In terms of your proposal:

1. It is not checked by default and haven't been for some time. Can't remember exactly which release this was changed, but it's been some time
2. I do like the idea of warning admins, but do not think it should be done when setting it for a client specifically. One thing I'd like to propose early next year is a built-in audit mechanism that can run through a set of policies and alert admins of bad configuration. By default we could have a policy enable to check for direct access grant; and admins will then get a warning if any clients leverage this. I'm also considering both a warning and an enforcement mode around this. More to come around this later though
3. I do like the idea of feature flagging ROPC. That is something we can introduce fairly easily
4. Related to 2 - my idea around security profile/policies there would be two built-in default policies. A lax one for development (start-dev) and a stricter default one for production (start). Rather than disabling ROPC by default I think we should work more on education/recommendations instead

**Reply by jsorah** [reply]
Yeah definitely agree with all that. I originally wrote this ~2 years ago, so certainly some things and thoughts for myself have changed. 

It is unrealistic to think the ROPC code itself would be dropped from Keycloak - too many legacy usages and use cases, and sometimes the risk is acceptable. Admins ultimately have to be able to make that decision.

My original intent with turning off ROPC by default as a feature (at least for new instances) was to just help strictly guide new admins/users of Keycloak towards best practices - its easy to ignore messages that say "Hey, this is a bad thing to do!" - if they _literally can't do it_ it might give them pause to think about what they are doing and _why_ that particular feature isn't so easily available. But I think in lieu of that, any auditing / security health kind of functionality like you mention in your 2 would be great.
